### ISSUE TITLE
Unable to load images for eval in tensorflow official resnet model

### ISSUE THREAD

**Issue by Spugnam** [tensorflow/models#4167]
### System information
- **What is the top-level directory of the model you are using**:
https://github.com/tensorflow/models/tree/master/official/resnet

- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
I have modified imagenet_main.py to change the hard-coded values related to the imagenet dataset. See code below.
I also used custom code to create TFRecords for my own dataset. I believe those files are good as the training of the resnet model seems to use the training images/ labels correctly.

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 
tf.GIT_VERSION/ tf.VERSION
v1.7.0-rc1-1088-g10878ede06/ 1.8.0-dev20180330
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**:
CUDA Version 9.1.85
cuDNN 7.1.3
- **GPU model and memory**:
2* GTX 1080i
- **Exact command to reproduce**:
cd models/official/resnet
python3 mydata_main.py \  # modified version of imagenet_main.py
--data_dir ~/path/to/TFRecords/ \
--model_dir /tmp/tests \
--export_dir /tmp/exports \
--train_epochs 2 \
--epochs_between_evals 1 \
--batch_size 256 \
--multi_gpu \
--hooks LoggingTensorHook \
--num_parallel_calls 12  \
--inter_op_parallelism_threads 0 \
--intra_op_parallelism_threads 0 \
--dtype fp32 \
--export_dir /tmp/resnet \
--version 1 \
--resnet_size 18

### Describe the problem
Note: I first created this [issue](https://stackoverflow.com/questions/50143417/unable-to-load-images-for-eval-in-tensorflow-official-resnet-model) on SO.
I'm trying to train the Tensorflow official resnet model on my own images and labels.
As explained above I changed imagenet_main.py to 'my_data_main.py'.
The images from my TFRecords are loaded correctly for the training but not for the evaluation. The following line in resnet_run_loop.py at the beginning of resnet_model_fn() loads the images in Tensorboard:

  tf.summary.image('images', features, max_outputs=6) 

I can see my images for the train run but not for eval in Tensorboard. See the SO issue for screenshots.

### Source code / logs

Beginning of mydata_main.py (replaces imagenet_main.py). I couldn't attach my (small) TFRecords so I'm afraid the issue can't be reproduced directly.

"""Runs a ResNet model on the ImageNet dataset."""                             

  from __future__ import absolute_import                                         
  from __future__ import division                                                
  from __future__ import print_function                                          

  import os                                                                      
  import sys                                                                     

  import tensorflow as tf  # pylint: disable=g-bad-import-order                  

  from official.resnet import imagenet_preprocessing                             
  from official.resnet import resnet_model                                       
  from official.resnet import resnet_run_loop                                    

  _DEFAULT_IMAGE_SIZE = 299                                                      
  _NUM_CHANNELS = 3                                                              
  _NUM_CLASSES = 9                                                               

  _NUM_IMAGES = {                                                                
      'train': 484,                                                              
      'validation': 121,                                                         
  }                                                                                 

  _NUM_TRAIN_FILES = 2                                                              
  _NUM_VAL_FILES = 2                                                                
  _SHUFFLE_BUFFER = 200                            

### Data processing                                                                 

  def get_filenames(is_training, data_dir):                                         
      """Return filenames for dataset."""                                           
      if is_training:                                                               
          return [                                                               
            os.path.join(data_dir, 'my_data_train_%05d-of-%05d.tfrecord' % (i, _NUM_TRAIN_FILES))
                for i in range(_NUM_TRAIN_FILES)]                                  
      else:                                                                      
          return [                                                               
            os.path.join(data_dir, 'my_data_validation_%05d-of-%05d.tfrecord' % (i, _NUM_VAL_FILES))
                for i in range(_NUM_VAL_FILES)] 

### rest of program unchanged

**Comment by k-w-w** [comment]
@robieta Can you help with this? Thanks!

**Comment by robieta** [comment]
Let me take a look.

**Comment by robieta** [comment]
I will admit this one threw me. It looks like this is an active feature request in TensorFlow. (https://github.com/tensorflow/tensorflow/issues/15332) It is possible to hack it in using hooks. (https://stackoverflow.com/questions/46852018/tensorflow-estimator-api-save-image-summary-in-eval-mode) Just note martinwicke's warnings about what a summary is in eval. Here is a simple example:

```
import tensorflow as tf


def input_fn(*args, **kwargs):
  batch_size = 20
  images = tf.zeros((batch_size, 10), tf.float32)
  labels = tf.zeros((batch_size, 1), tf.int32)
  return tf.data.Dataset.from_tensors((images, labels)).repeat(10)

def model_fn(features, labels, mode, params):
  pred = tf.layers.dense(inputs=features, units=1)

  if mode == tf.estimator.ModeKeys.TRAIN:
    loss = tf.losses.absolute_difference(labels, pred)
    tf.summary.scalar(name="train_scalar", tensor=1)
    optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.1)
    return tf.estimator.EstimatorSpec(
        mode=tf.estimator.ModeKeys.TRAIN,
        loss=loss,
        train_op=optimizer.minimize(loss, tf.train.get_or_create_global_step()))
  if mode == tf.estimator.ModeKeys.EVAL:
    loss = tf.losses.absolute_difference(labels, pred)
    tf.summary.scalar(name="eval_scalar", tensor=2)

    # Use a hook to force evaluation. If commented out the scalar will not be saved.
    eval_summary_hook = tf.train.SummarySaverHook(
        save_steps=1,
        output_dir= "/tmp/summary_test/eval_core",
        summary_op=tf.summary.merge_all())

    return tf.estimator.EstimatorSpec(
        mode=tf.estimator.ModeKeys.EVAL,
        loss=loss,
        evaluation_hooks=[eval_summary_hook]
    )


def main(_):
  model_dir = "/tmp/summary_test"
  if tf.gfile.Exists(model_dir):
    tf.gfile.DeleteRecursively(model_dir)

  run_config = tf.estimator.RunConfig(save_summary_steps=1)

  estimator = tf.estimator.Estimator(
      model_fn=model_fn,
      model_dir=model_dir,
      params={},
      config=run_config
  )

  for _ in range(10):
    estimator.train(input_fn=input_fn)
    estimator.evaluate(input_fn=input_fn)

if __name__ == "__main__":
  main([])
```

Hope this helps, and once evaluate has better support for summaries I will definitely look to save more evaluation info by default.

**Comment by Spugnam** [comment]
Thanks a ton, I'll look at the links.

**Comment by Spugnam** [comment]
I was able to display some evaluation images in Tensorboard using the SO answer thanks. I was only able to make it work on the original imagenet data though. I still have the problem with my own data. More importantly (although it might be the same root cause?) I can't get an evaluation accuracy or loss, they are all 0. See my run log below:

INFO:tensorflow:Starting evaluation at 2018-05-08-21:59:35
INFO:tensorflow:Graph was finalized.
2018-05-08 17:59:35.425106: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1435] Adding visible gpu devices: 0, 1
2018-05-08 17:59:35.425178: I tensorflow/core/common_runtime/gpu/gpu_device.cc:923] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-05-08 17:59:35.425186: I tensorflow/core/common_runtime/gpu/gpu_device.cc:929]      0 1 
2018-05-08 17:59:35.425192: I tensorflow/core/common_runtime/gpu/gpu_device.cc:942] 0:   N Y 
2018-05-08 17:59:35.425197: I tensorflow/core/common_runtime/gpu/gpu_device.cc:942] 1:   Y N 
2018-05-08 17:59:35.425359: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1053] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10209 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:03:00.0, compute capability: 6.1)
2018-05-08 17:59:35.425444: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1053] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:1 with 10207 MB memory) -> physical GPU (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:04:00.0, compute capability: 6.1)
INFO:tensorflow:Restoring parameters from /mnt/sda1/tensorflow_logs/popsy_resnet/model.ckpt-15
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2018-05-08-21:59:35
INFO:tensorflow:Saving dict for global step 15: accuracy = 0.0, global_step = 15, loss = 0.0

I am running out of ideas since:
- I only changed a few parameters in imagenet_main.py to point to my .tfrecords files
- my tfrecords are read properly on the training side

hope I'm missing something simple?
appreciate any help

**Comment by Spugnam** [comment]
For background here's the code I added in resnet_run_loop.py to visualize eval images: 

  306     accuracy = tf.metrics.accuracy(                                            
  307         tf.argmax(labels, axis=1), predictions['classes'])                     
  308     metrics = {'accuracy': accuracy} 

# my addition
  312     if mode == tf.estimator.ModeKeys.EVAL:                                     
  313         tf.identity(accuracy[1], name='eval_accuracy')                         
  314         tf.summary.scalar("eval_accuracy", accuracy[1])                        
  315                                                                                
  316         # tensorflow issue 4167                                                
  317         # Use a hook to force evaluation. If commented out the scalar will not be saved.
  318         eval_summary_hook = tf.train.SummarySaverHook(                         
  319             save_steps=1,                                                      
  320             # Per SO eval folder needs to be 1 level below tensorboard logdir (model_dir)
  321             output_dir="/mnt/sda1/tensorflow_logs/mydata_resnet/eval_core",     
  322             summary_op=tf.summary.merge_all())                                 
  323                                                                                
  324         return tf.estimator.EstimatorSpec(                                     
  325             mode=mode,                                                         
  326             predictions=predictions,                                           
  327             loss=loss,                                                         
  328             train_op=None,                                                     
  329             evaluation_hooks=[eval_summary_hook],                              
  330             eval_metric_ops=metrics)

**Comment by robieta** [comment]
For this I'm going to have to refer you to the tensorflow [Stack Overflow](http://stackoverflow.com/questions/tagged/tensorflow).

**Comment by tabrisweapon** [comment]
Same problem here.
I reorganize the official resnet codes to make it more succinct, but did not change anything aggressive. The training process works well, however, the evaluation process gets 0 loss and 0 accuracy. Any ideas?
